File: lab1-22.02.21/src/main/java/testpassword/lab1/controllers/AdvertController.java
package testpassword.lab1.controllers;

import lombok.val;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;
import testpassword.lab1.requests.AdvertReq;
import testpassword.lab1.responses.AdvertRes;
import testpassword.lab1.security.JWTUtil;
import testpassword.lab1.services.AdvertService;
import testpassword.lab1.services.UserService;
import javax.servlet.http.HttpServletRequest;
import javax.validation.Valid;
import java.util.Arrays;
import java.util.Objects;
import java.util.function.Predicate;
import java.util.stream.Collectors;

@RestController @RequestMapping(path = "/advert")
public class AdvertController {

    @Autowired private JWTUtil jwtUtil;
    @Autowired private UserService userService;
    @Autowired private AdvertService advertService;

    @GetMapping(path = "getAll", produces = "application/json")
    public ResponseEntity<AdvertRes> getAll() {
        return new ResponseEntity<>(AdvertRes.builder().adverts(advertService.getAll()).build(), HttpStatus.OK);
    }

    @GetMapping(path = "get/{userId}", produces = "application/json")
    public ResponseEntity<AdvertRes> getForUser(@PathVariable long userId) {
        return userService.get(userId)
                .map(u -> new ResponseEntity<>(
                        AdvertRes.builder().adverts(u.getAdverts()).build(),
                        HttpStatus.OK)
                ).orElse(new ResponseEntity<>(
                        AdvertRes.builder().msg("Can't find user with requested id").build(),
                        HttpStatus.BAD_REQUEST));
    }

    @PostMapping(path = "getFiltered", consumes = "application/json", produces = "application/json")
    public ResponseEntity<AdvertRes> get(@Valid @RequestBody AdvertReq req) {
        return new ResponseEntity<>(
                AdvertRes.builder().adverts(
                        Arrays.stream(req.advertsIds)
                                .mapToObj(advertService::get)
                                .filter(Objects::nonNull)
                                .collect(Collectors.toSet()))
                        .build(),
                HttpStatus.ACCEPTED);
    }

    @PutMapping(path = "add", consumes = "application/json", produces = "application/json")
    public ResponseEntity<AdvertRes> add(@Valid @RequestBody AdvertReq req, HttpServletRequest rawReq) {
        return (userService.get(jwtUtil.decode(rawReq)))
                .map(u -> {
                    val advert = req.toAdvert();
                    Predicate<String> badField = x -> x == null || x.equals("");
                    if (badField.test(advert.getMobileNumber()) || badField.test(advert.getName()))
                        return new ResponseEntity<>(
                                AdvertRes.builder().msg("Use should set mobile phone or name").build(),
                                HttpStatus.BAD_REQUEST);
                    advert.setUser(u);
                    return (advertService.add(advert)) ?
                            new ResponseEntity<>(
                                    AdvertRes.builder().msg("Your advert was publish!").build(),
                                    HttpStatus.CREATED) :
                            new ResponseEntity<>(
                                    AdvertRes.builder().msg("We found a problems while moderating. Please read out rules").build(),
                                    HttpStatus.I_AM_A_TEAPOT);
                        }
                ).orElse(new ResponseEntity<>(AdvertRes.builder().msg("User didn't exist").build(), HttpStatus.UNAUTHORIZED));
    }

    @DeleteMapping(path = "delete", consumes = "application/json", produces = "application/json")
    public ResponseEntity<AdvertRes> delete(@Valid @RequestBody AdvertReq req, HttpServletRequest rawReq) {
        Arrays.stream(req.getAdvertsIds())
                .mapToObj(advertService::get)
                .filter(a -> a.getUser().getUserId() == userService.get(jwtUtil.decode(rawReq)).get().getUserId())
                .forEach(advertService::delete);
        return new ResponseEntity<>(AdvertRes.builder().msg("Successfully deleted").build(), HttpStatus.OK);
    }
}

File: lab3-18.05.21/src/main/kotlin/lab3/controllers/AdvertController.kt
package lab3.controllers

import lab3.dtos.http.AdvertReq
import lab3.dtos.http.AdvertRes
import lab3.models.Advert
import lab3.security.JWTTokenUtil
import lab3.services.AdvertService
import lab3.services.UserService
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.http.*
import org.springframework.web.bind.annotation.*
import javax.servlet.http.HttpServletRequest
import javax.validation.Valid

@RestController @RequestMapping(path = ["/advert"])
class AdvertController {

    @Autowired private lateinit var jwt: JWTTokenUtil
    @Autowired private lateinit var userService: UserService
    @Autowired private lateinit var advertService: AdvertService

    private fun ok(block: AdvertRes.() -> Unit) = ResponseEntity(AdvertRes().apply(block), HttpStatus.OK)

    @GetMapping(path = ["all"], produces = ["application/json"])
    fun getAll() = ok { adverts = advertService.getAll() }

    @GetMapping(path = ["{advertId}"], produces = ["application/json"])
    fun getAdvert(@PathVariable advertId: Long) = ok { adverts = listOf(advertService[advertId]) }

    @GetMapping(path = ["forUser/{userId}"], produces = ["application/json"])
    fun getForUser(@PathVariable userId: Long) =
        ok {
            adverts = userService[userId].adverts
        }

    @GetMapping(path=["spec"], produces = ["application/json"])
    fun getSpecificAdverts(@RequestParam ids: String) =
        ok {
            adverts = ids.split(";").mapNotNull {
                try {
                    advertService[it.toLong()]
                } catch (e: Exception) {
                    null
                }
            }
        }

    @PostMapping(path = ["add"], consumes = ["application/json"], produces = ["application/json"])
    fun add(@Valid @RequestBody req: AdvertReq, raw: HttpServletRequest) =
        ok {
            println("a norm $req")
            if (req.mobileNumber.isNullOrBlank() || req.name.isNullOrBlank()) throw Exception("Required mobileNumber and name for advert")
            val advert = req.fillAdvert(Advert())
            advert.user = userService loadUserByUsername (jwt decode raw)
            msg = if (advertService add advert) "Your advert was complete automoderation and sent to manual"
            else "We found a problems while moderating. Please read our rules"
        }

    @PutMapping(path = ["{advertId}"], consumes = ["application/json"], produces = ["application/json"])
    fun modify(@PathVariable advertId: Long, @RequestBody req: AdvertReq, raw: HttpServletRequest) =
        ok {
            userService loadUserByUsername (jwt decode raw)
            advertService.modify(advertId, req)
            msg = "Successfully modified"
        }

    @DeleteMapping(path = ["delete"], consumes = ["application/json"], produces = ["application/json"])
    fun delete(@RequestParam ids: String, @Valid @RequestBody req: AdvertReq, raw: HttpServletRequest) =
        ok {
            ids.split(";")
                .map { advertService[it.toLong()] }
                .filter { it.user.userId == (userService loadUserByUsername(jwt decode raw)).userId }
                .forEach(advertService::delete)
            msg = "Successfully deleted"
        }
}